libgraphics/src/drawing.cpp

```diff
@@ -0,0 +1,1894 @@
+            }
+            break;
+                
+            case kMCGDrawingPathOpcodeQuadraticTo:
+            case kMCGDrawingPathOpcodeRelativeQuadraticTo:
+            {
+                MCGFloat t_ax, t_ay, t_x, t_y;
+                if (!Scalar(t_ax) || !Scalar(t_ay) ||
+                    !Scalar(t_x) || !Scalar(t_y))
+                {
+                    break;
+                }
+                p_visitor.PathQuadraticTo(t_opcode == kMCGDrawingPathOpcodeRelativeQuadraticTo, t_ax, t_ay, t_x, t_y);
+            }
+            break;
+                
+            case kMCGDrawingPathOpcodeSmoothQuadraticTo:
+            case kMCGDrawingPathOpcodeRelativeSmoothQuadraticTo:
+            {
+                MCGFloat t_x, t_y;
+                if (!Scalar(t_x) || !Scalar(t_y))
+                {
+                    break;
+                }
+                p_visitor.PathSmoothQuadraticTo(t_opcode == kMCGDrawingPathOpcodeRelativeSmoothQuadraticTo, t_x, t_y);
+            }
+            break;
+                
+            case kMCGDrawingPathOpcodeArcTo:
+            case kMCGDrawingPathOpcodeRelativeArcTo:
+            case kMCGDrawingPathOpcodeReflexArcTo:
+            case kMCGDrawingPathOpcodeRelativeReflexArcTo:
+            case kMCGDrawingPathOpcodeReverseArcTo:
+            case kMCGDrawingPathOpcodeRelativeReverseArcTo:
+            case kMCGDrawingPathOpcodeReverseReflexArcTo:
+            case kMCGDrawingPathOpcodeRelativeReverseReflexArcTo:
+            {
+                MCGFloat t_rx, t_ry, t_rotation, t_x, t_y;
+                if (!LengthScalar(t_rx) || !LengthScalar(t_ry) ||
+                    !AngleScalar(t_rotation) ||
+                    !Scalar(t_x) || !Scalar(t_y))
+                {
+                    break;
+                }
+                
+                bool t_is_relative =
+                            t_opcode == kMCGDrawingPathOpcodeRelativeArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeRelativeReflexArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeRelativeReverseArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeRelativeReverseReflexArcTo;
+                bool t_is_reflex =
+                            t_opcode == kMCGDrawingPathOpcodeReflexArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeRelativeReflexArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeReverseReflexArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeRelativeReverseReflexArcTo;
+                bool t_is_reverse =
+                            t_opcode == kMCGDrawingPathOpcodeReverseArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeRelativeReverseArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeReverseReflexArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeRelativeReverseReflexArcTo;
```

Comment:
Again totally inessential, but you've got inlined functions (eg `MCGDrawingPreserveAspectRatioIsYMid`) for this sort of thing elsewhere, so might be worth doing for these

Reply:
Sure - that's a good idea.